src/Provider.js

```diff
@@ -1,42 +1,37 @@
-import { Children, Component, createContext, createElement } from "react"
+import React, { createContext, useContext, useRef } from "react"
 import { shallowEqual } from "./utils/utils"
 
 const specialReactKeys = { children: true, key: true, ref: true }
 
 export const MobXProviderContext = createContext({})
 
-export class Provider extends Component {
-    static contextType = MobXProviderContext
-    static displayName = "MobXProvider"
-
-    constructor(props, context) {
-        super(props, context)
-        this.state = {
-            ...context,
+export function Provider(props) {
+    const parentValue = useContext(MobXProviderContext)
+    const valueRef = useRef(null)
+    if (valueRef.current === null) {
+        valueRef.current = {
+            ...parentValue,
             ...grabStores(props)
         }
     }
-
-    render() {
-        return createElement(
-            MobXProviderContext.Provider,
-            { value: this.state },
-            this.props.children
-        )
-    }
-
-    static getDerivedStateFromProps(nextProps, prevState) {
-        if (process.env.NODE_ENV !== "production") {
-            const newStores = { ...prevState, ...grabStores(nextProps) } // spread in prevState for the context based stores
-            if (!shallowEqual(prevState, newStores))
-                throw new Error(
-                    "MobX Provider: The set of provided stores has changed. Please avoid changing stores as the change might not propagate to all children"
-                )
+    const value = valueRef.current
+
+    if (process.env.NODE_ENV !== "production") {
+        const newValue = { ...value, ...grabStores(props) } // spread in previous state for the context based stores
+        if (!shallowEqual(value, newValue)) {
+            throw new Error(
+                "MobX Provider: The set of provided stores has changed. Please avoid changing stores as the change might not propagate to all children"
```

Comment:
I am thinking if this is still valid actually. Isn't it from legacy context too? I think it's because the legacy context could not propagate to pure components. The Context does not have that problem. I don't see a reason why the value could not change. I mean in MobX world it doesn't usually need to, but people are crazy sometimes :)

Reply:
The quote from the change log for `mobx-react@6.0.0`:
> * Changing the set of stores in `Provider` is no longer supported and while throw a hard error (this was a warning before), as the model of `Provider` / `inject` has always been to inject final values into the tree. (That is, fixed references, injected objects themselves can be stateful without problem). If you want to dynamically swap what is provided into the tree, use `React.createContext` instead of `Provider` / `inject`. The  suppressChangedStoreWarning` flag for `Provider` has been dropped.

https://github.com/mobxjs/mobx-react/commit/a64c2d365300f669463c0f7d9fcde42337207807#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR11

Reply:
Probably this can be discussed separately from this PR so as not to delay the merge. What do you think?

Reply:
I think we should ask @mweststrate about this change in 6.0.0.

Reply:
I don't think we need a new issue considering this PR is not bug fixing there is no need to hurry.

Presently, I don't see a reason why this should be restricted.

Reply:
I prefer more granular PRs and branches that have one clear goal. It has some connotations with The Single Responsibility Principle, for example. In this PR I want to rewrite the Provider component to function component, so it has not nothing to do with this restriction.
It has the following benefits:
1) the PR will not hung
2) it will be more simple to investigate when and why some changes were made in the future.

Reply:
Well, in that case, would you mind adding some tests for the Provider? If I see correctly, we have none. In theory, this change shouldn't break anything, but better to be sure than sorry. Perhaps even add those tests in a separate PR to see them passing and then merge here.

Reply:
Thanks. Yes, I will look into it.

Reply:
We have some tests in `inject.test.js`. I will move them in a separate file,  review them, refactor them if necessary and probably write new ones.

Reply:
> I don't think we need a new issue considering this PR is not bug fixing so there is no need to hurry.

https://github.com/mobxjs/mobx-react/issues/745

Reply:
@FredyC This PR is ready to merge.